extensions/narayana-jta/runtime/src/main/java/io/quarkus/narayana/jta/runtime/NarayanaJtaRecorder.java

```diff
@@ -158,7 +157,12 @@ public void startRecoveryService(final TransactionManagerConfiguration transacti
     public void handleShutdown(ShutdownContext context, TransactionManagerConfiguration transactions) {
         context.addLastShutdownTask(() -> {
             if (transactions.enableRecovery) {
-                RecoveryManager.manager().terminate(true);
+                try {
+                    QuarkusRecoveryService.getInstance().stop();
+                } catch (Exception e) {
+                    throw new RuntimeException(e);
+                }
+                QuarkusRecoveryService.getInstance().destroy();
```

Comment:
Shouldn't this be part of a finally clause?

I'm a bit worried of the state we will end up in if something wrong happens when we stop the service.

Reply:
Good call on both counts.

The Narayana RecoveryManager#terminate call [1] checks the state [2] and throws IllegalStateException if it has previously been shutdown so it's really a sanity check telling the caller not to call it twice.

[1] https://github.com/jbosstm/narayana/blob/6.0.1.Final/ArjunaCore/arjuna/classes/com/arjuna/ats/arjuna/recovery/RecoveryManager.java#L206
[2] https://github.com/jbosstm/narayana/blob/6.0.1.Final/ArjunaCore/arjuna/classes/com/arjuna/ats/arjuna/recovery/RecoveryManager.java#L489